### ISSUE TITLE
Child class overrides with a different return type are not detected

### ISSUE THREAD

**Issue by Kjir** [microsoft/pyright#3193]
**Description**

If a base class is extended and a method `m` is overridden with a different return type, all functions from the base class that return the result calling `self.m` have the wrong signature type - the one coming from the base class.

**Reproduction**
This code shows the problem:

```python
import asyncio
from typing_extensions import reveal_type


class Base:
    def salute(self):
        return self.sayHi("Base")

    def sayHi(self, name):
        return f"Hello, {name}"


class AsyncChild(Base):
    async def sayHi(self, name):
        await asyncio.sleep(.1)
        return f"Hello, {name}"


b = Base()
reveal_type(b.salute) # Type of "b.salute" is "() -> str"
reveal_type(b.sayHi) # Type of "b.sayHi" is "(name: Unknown) -> str"

ac = AsyncChild()
reveal_type(ac.salute) # Type of "ac.salute" is "() -> str"   <--- This is wrong!
reveal_type(ac.sayHi) # Type of "ac.sayHi" is "(name: Unknown) -> Coroutine[Any, Any, str]"  <--- This is correct
```
The inferred return type is taken from the base class, but it has been changed by the overriding in the client class. The suggested typing does not correspond to what will actually be returned when running this (valid) Python code.

**Expected behavior**
Adding the implementation of `salute` to the child class shows the correct type:

```python
import asyncio
from typing_extensions import reveal_type


class Base:
    def salute(self):
        return self.sayHi("Base")

    def sayHi(self, name):
        return f"Hello, {name}"


class AsyncChild(Base):
    def salute(self):
        return self.sayHi("Base")

    async def sayHi(self, name):
        await asyncio.sleep(0.1)
        return f"Hello, {name}"


b = Base()
reveal_type(b.salute) # Type of "b.salute" is "() -> str"

ac = AsyncChild()
reveal_type(ac.salute) # Type of "ac.salute" is "() -> Coroutine[Any, Any, str]"
```

It should not be necessary to reimplement the `salute` method to infer the correct return type.

**Version**
The installed `pyright` version is `1.1.229`.

**Comment by erictraut** [comment]
Yes, this is expected behavior. The inferred return type of a function is based on the information available within that class. 

You can provide return type annotations if the inference is producing results that are different from your intent.

**Comment by erictraut** [comment]
I'll also point out that it's an anti-pattern to override a method within a child class in a manner that is incompatible with the base class method of the same name. This results in extremely fragile code. Most languages wouldn't even allow this, but Python is a bit lax because it's such a dynamic language.

Pyright will detect and report this condition if you enable the diagnostic check called `reportIncompatibleMethodOverride`.

**Comment by Kjir** [comment]
> Yes, this is expected behavior. The inferred return type of a function is based on the information available within that class.

It's by design that the type suggestion is wrong? It is **not** what you get at runtime and the type system will in fact complain if you do the right thing.

I would argue that the role of type system is to warn you of such possible mistakes, not lead you into them. If the type can't be figured out, it should be `Unknown`, not the wrong type. Imagine introducing `pyright` to a codebase that already has such a pattern: it's either going to be a battle of man vs typing system in the good scenario, or the unfortunate developer will adapt all call sites to comply with the type suggestion introducing bugs. 

> You can provide return type annotations if the inference is producing results that are different from your intent.

That works, but can be extremely inconvenient if you have to do it for a large number of methods.

> I'll also point out that it's an anti-pattern to override a method within a child class in a manner that is incompatible with the base class method of the same name.

Maybe yes, maybe not. I would say it depends. If you do not intend to use the base class ever it is a reasonable approach.

The way I stumbled on this is such an example: we use a vendor's library that implements a client for their API. It will do requests synchronously, but we want to leverage Twisted (or asyncio, the details don't matter). All the methods are implemented with such a pattern:

```python
    def create_group(self, request, group_id=None):
        return self.start().uri('/api/group') \
            .url_segment(group_id) \
            .body_handler(JSONBodyHandler(request)) \
            .post() \
            .go()
```

The last method call `go()` would perform the request that was built, in what essentially is a builder pattern. The `start()` method is where the builder is created:

```python
    def start(self):
        client = RESTClient().url(self.base_url)
        if self.tenant_id is not None:
            client.header("X-TenantId", self.tenant_id)

        return client
```

So to enable async behaviour in the library all we had to do was reimplement the `start` method and the `go` method of the `RESTClient`. Yes, the returned type changes, no, there is no source of confusion since we only use our wrapper in our project. Not ideal? Maybe, but reimplementing the client completely (or copy/paste its implementation) is not a very good use of our time and it would make the upgrade path substantially impossible. The only tradeoff is that the return type changes and the hope was that having tooling like `pyright` would help clarify this change.

**Comment by erictraut** [comment]
Type checkers need to make some assumptions — especially when it comes to type inference. You're violating some of these assumptions by using fragile and non-recommended practices within your code. You'll find that other type checkers (both in Python and in other languages) make the same assumption with respect to method overrides.

I understand your justification for your use of an incompatible override, but it is still atypical, fragile and not recommended. If you are going to violate assumptions like this, you can inform type checkers of your intent by adding type annotations.

**Comment by Kjir** [comment]
Do you have a link to those recommendations?
`mypy` does provide an equivalent warning if enabled, but it will not provide a wrong type hint. It will suggest `Any` and while that is not a very helpful hint, at least it is not wrong and misleading.

**Comment by Pyprohly** [comment]
The recommendation is that you don’t override a method with a different signature, and this includes the return type.

When you write `async def f() -> str` its like sugar for `def f() -> Coroutine[Optional[Future[None]], None, str]`. The type `str` and `Coroutine` are different and so overriding a sync function with an async one is never compatible in a strongly typed setting.

It looks like you’re trying to make an async and sync IO compatible library. If you try too hard to avoid code duplication, things get messy very quickly. Just allow the duplication to happen. It will be neater. If you need a base class try to avoid putting methods on it, only fields if you have to.

> If the type can't be figured out, it should be Unknown, not the wrong type.

> [Mypy] will suggest Any and while that is not a very helpful hint, at least it is not wrong and misleading.

If there are typing errors in your code then you can’t rely on `reveal_type` giving accurate information. As was suggested, enable `reportIncompatibleMethodOverride`.

**Comment by Kjir** [comment]
> The recommendation is that you don’t override a method with a different signature, and this includes the return

I still haven't seen where this recommendation comes from. First of all a recommendation is not mandatory to be followed, and secondly this is perfectly valid Python code, and for perfectly valid Python code `pyright` is giving a wrong answer.

I suspect you are talking about the Liskov Substitution Principle, which as everything it is not free from criticism, and most importantly it is completely irrelevant in this case since the base class is not used anywhere except as a base. There is no need for substitution ever.

> It looks like you’re trying to make an async and sync IO compatible library. 

That is exactly what I'm doing, as stated above

> If you try too hard to avoid code duplication, things get messy very quickly. Just allow the duplication to happen. It will be neater. If you need a base class try to avoid putting methods on it, only fields if you have to.

Thanks for the suggestion. Unfortunately, as stated above, I'm extending a third-party library, I have no control over that. Furthermore we don't have the resources to maintain that library - we aren't Microsoft - especially since the current solution works perfectly and the only downside I've seen stated so far is that it is "not recommended" without any real justification.

It is also not true that allowing the duplication to happen will make things "neater", nor that it will "get messy very quickly". Only one method needed to change and nothing more. In fact by allowing the duplication to happen I would not only go from about 60 lines of code to over 4000 lines of code. This is not neater.

While in general I agree that changing the behaviour by composition rather than by inheritance would be preferable, it is still a valid approach.

> If there are typing errors in your code then you can’t rely on `reveal_type` giving accurate information. As was suggested, enable `reportIncompatibleMethodOverride`.

If it's a typing error, then the `reportIncompatibleMethodOverride` options should be one by default, as it is in Typescript. It would be way better than getting the wrong type suggestion because it would make `pyright` more dependable, whereas as it is now I have trouble believing what it is telling me. What else does the tool get wrong because there's code that follows not recommended patterns?

**Comment by Pyprohly** [comment]
The bottom line is that your code is not inherently type-safe. The type inference doesn’t align with the runtime behaviour because your code doesn’t make sense in a type-safe setting, even if it works fine at runtime. Overriding a function in the manner you’ve demonstrated would fail to compile in a statically typed language. Pyright does reveal the error, you just have to configure it.

Keeping the sync and async worlds as separate as possible is the advice I would give from my experience building a library that supports both sync and async. The implications may be different for you, extending a third-party library that only supports one IO model into both sync and async worlds, I wouldn’t know.

To avoid being surprised by the type inference rules, I strongly advise you to annotate your functions and methods. This is something you should always do. Also, as you indicated, I believe composition would substantially benefit your codebase. Create an adaptor class for the third-party class that has to be extended so that your `Any` annotations and `type: ignore`s are all in one place.

**Comment by Kjir** [comment]
> Overriding a function in the manner you’ve demonstrated would fail to compile in a statically typed language

Except Python is not a statically typed language :shrug: 

No matter, I get the point. I still don't know why you don't turn on the functionality by default - since the goal seems to create a static language with different rules than Python's - nor why you think that giving a wrong hint is acceptable.

I will work around the tool, I already added a comment in the code to explain why the boilerplate is needed.
